<a href="https://colab.research.google.com/github/srishtiarora42006-hue/Answer/blob/main/ADVANCED_PANDAS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Q1 Dataset Exploration

Perform the following exploratory steps:

Display the first 10 rows

Display the dataset structure using info()

Find the number of rows and columns

Identify the data types of each column

In [1]:
import pandas as pd

In [3]:
pd.read_excel("/content/Advanced_pandas_sales_dataset.xlsx")

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
...,...,...,...,...,...,...,...,...,...,...,...
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00
226,1227,2024-04-02,Dev,Pune,South,Mouse,Accessories,1669,6,0.05,323.13
227,1228,2024-01-14,Vikas,Chennai,North,Tablet,Electronics,2682,7,0.15,119.82
228,1229,2024-09-04,Priya,Pune,East,Tablet,Electronics,2635,8,0.00,531.76


In [4]:
df = pd.read_excel("/content/Advanced_pandas_sales_dataset.xlsx")

In [5]:
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Order_ID    230 non-null    int64         
 1   Order_Date  230 non-null    datetime64[ns]
 2   Customer    230 non-null    object        
 3   City        230 non-null    object        
 4   Region      230 non-null    object        
 5   Product     230 non-null    object        
 6   Category    230 non-null    object        
 7   Sales       230 non-null    int64         
 8   Quantity    230 non-null    int64         
 9   Discount    230 non-null    float64       
 10  Profit      230 non-null    float64       
dtypes: datetime64[ns](1), float64(2), int64(3), object(5)
memory usage: 19.9+ KB


In [7]:
df.shape

(230, 11)

In [9]:
df.dtypes

,0
Order_ID,int64
Order_Date,datetime64[ns]
Customer,object
City,object
Region,object
Product,object
Category,object
Sales,int64
Quantity,int64
Discount,float64


**Q2** Sales Filtering Analysis

Filter the dataset to display:

Orders where Sales > 3500

Orders where Region = "East" and Sales > 2500

Explain what insights you observe.

In [11]:
df[df["Sales"] > 3500]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
10,1011,2024-06-30,Neha,Delhi,South,Printer,Office,5977,6,0.00,1578.32
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [16]:
df[(df["Sales"] >2500 ) & (df["Region"] == "East")]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
34,1035,2024-02-20,Nitin,Ahmedabad,East,Laptop,Electronics,2535,4,0.10,188.15
41,1042,2024-07-15,Rohan,Chennai,East,Laptop,Electronics,3268,8,0.15,295.15
43,1044,2024-09-20,Kavya,Hyderabad,East,Printer,Office,2618,7,0.05,454.64


Insights observed from the above data were:
- East region has a relatively higher number of sales which are above 2500.
- it can also be observed that overall from all the regions 50% of the orders are above the order value 3500
- The highest order value from the above data was ₹5986.

**Q3** Customer Purchase Analysis

Identify the top 5 customers based on total sales.

Hint: Use GroupBy with aggregation.

In [20]:
# top 5 customers
df.groupby("Customer")['Sales'].sum().sort_values(ascending=False).head(5)

,Sales
Customer,
Nitin,59820
Raj,58416
Rahul,52291
Dev,48500
Isha,46818


**Q4** Regional Sales Performance

Calculate the total sales generated by each region.

Then identify:

Which region generated the highest revenue?

In [21]:
Regional_sales = df.groupby("Region")['Sales'].sum()

In [22]:
print(Regional_sales)

Region
East     232675
North    158210
South    183108
West     170227
Name: Sales, dtype: int64


In [24]:
Highest_revenue = Regional_sales.idxmax()
print("The region with the highest revenue is:", Highest_revenue)

The region with the highest revenue is: East


**Q5** Product Performance Analysis

Calculate the total quantity sold for each product.

Which product appears to be the most popular?

In [27]:
Total_quantity_sold = df.groupby("Product")['Quantity'].sum()
print(Total_quantity_sold)

Product
Headphones    115
Keyboard      110
Laptop        140
Monitor       156
Mouse         122
Phone         151
Printer       108
Tablet        150
Name: Quantity, dtype: int64


In [28]:
Most_popular_product = Total_quantity_sold.idxmax()
print("The most popular product is:", Most_popular_product)

The most popular product is: Monitor


**Q6**Multi-Level Grouping

Group the dataset by: Region and Category

Calculate the total sales for each combination.

What insights can you derive from this analysis?

In [29]:
#group by region and category
df.groupby("Region")["Category"].value_counts()

Region  Category   
East    Electronics    48
        Accessories    18
        Office          6
North   Electronics    23
        Accessories    22
        Office          8
South   Electronics    26
        Accessories    24
        Office          4
West    Electronics    26
        Accessories    17
        Office          8
Name: count, dtype: int64

In [30]:
#total sales for each category
total_sales = df.groupby(["Region", "Category"])["Sales"].sum()
print(total_sales)

Region  Category   
East    Accessories     57646
        Electronics    160907
        Office          14122
North   Accessories     62052
        Electronics     70023
        Office          26135
South   Accessories     80914
        Electronics     86690
        Office          15504
West    Accessories     64811
        Electronics     84128
        Office          21288
Name: Sales, dtype: int64


Insights derived from the above analysis are:-
- The maximum sales of accessories and Electronics were in the East region.
- The maximum sales of office supplies were in North region.


**Q7** Pivot Table Creation

Create a pivot table showing:

Rows → Region

Columns → Product

Values → Total Sales

Which region sells the most laptops?

In [31]:
pd.pivot_table(df, index="Region", columns="Product", values="Sales", aggfunc="sum")

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


The East region sells the most number of laptops.

**Q8** Profitability Analysis

Calculate the average profit for each product category.

Which category is most profitable?



In [34]:
average_sales = df.groupby("Category")["Profit"].mean()
print(average_sales)

Category
Accessories    410.603086
Electronics    396.462114
Office         471.938462
Name: Profit, dtype: float64


 Office supplies category is the most profitable

**Q9** Discount Impact Study

Analyze the relationship between discount and profit.

Tasks:

Calculate the average profit for each discount level

Identify whether higher discounts lead to lower profit

In [35]:
#average profit for each discount level
df.groupby("Discount")["Profit"].mean()

,Profit
Discount,
0.00,751.210588
0.05,531.368039
0.10,346.460513
0.15,237.191400
0.20,90.026923


As per the insight derived we can observe that higher rates of discounts lead to decreasing profits, hence discounts levied  and average profits earned have an inverse relationship.


**Q10**End-to-End Data Wrangling Task

Perform the following workflow:

Filter the dataset where Sales > 3000

Group data by Region and Product

Calculate total sales

Reshape the result into a pivot table

Explain the business insights obtained from this analysis.

In [36]:
#filter the data
df[df["Sales"]>3000]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [43]:
#group data by region and product
df.groupby(["Region" , "Product"])["Sales"].sum()


Region  Product   
East    Headphones    23899
        Keyboard      10845
        Laptop        55568
        Monitor       31299
        Mouse         22902
        Phone         39667
        Printer       14122
        Tablet        34373
North   Headphones    15450
        Keyboard      20538
        Laptop         4540
        Monitor       14922
        Mouse         26064
        Phone         30535
        Printer       26135
        Tablet        20026
South   Headphones    24111
        Keyboard      23480
        Laptop        23029
        Monitor       19167
        Mouse         33323
        Phone         18218
        Printer       15504
        Tablet        26276
West    Headphones    17159
        Keyboard      38499
        Laptop        10826
        Monitor       18178
        Mouse          9153
        Phone         27850
        Printer       21288
        Tablet        27274
Name: Sales, dtype: int64

In [44]:
# pivot table

pd.pivot_table(df, index="Region", columns="Product", values="Sales", aggfunc="sum")

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


Insights observed obtained from this analysis show that  the East region generated the highest sales for Laptops (₹55,568), making it the strongest region-product combination. The West region performed particularly well for Keyboards (₹38,499), while the South region had the highest sales for Mice (₹33,323). Overall, the results show that product performance varies across regions, which can help the business plan inventory and marketing strategies according to regional demand.
